cmd/soroban-rpc/internal/db/event.go

```diff
@@ -0,0 +1,336 @@
+}
+
+//nolint:gocognit,funlen,cyclop
+func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+	txCount := lcm.CountTransactions()
+
+	if eventHandler.stmtCache == nil {
+		return errors.New("EventWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	var txReader *ingest.LedgerTransactionReader
+	txReader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to open transaction reader for ledger %d: %w ",
+			lcm.LedgerSequence(), err)
+	}
+	defer func() {
+		closeErr := txReader.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	for {
+		var tx ingest.LedgerTransaction
+		tx, err = txReader.Read()
+		if errors.Is(err, io.EOF) {
+			err = nil
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if !tx.Result.Successful() {
+			continue
+		}
+
+		transactionHash := tx.Result.TransactionHash[:]
+
+		txEvents, err := tx.GetDiagnosticEvents()
+		if err != nil {
+			return err
+		}
+
+		if len(txEvents) == 0 {
+			continue
+		}
+
+		query := sq.Insert(eventTableName).
+			Columns(
+				"id",
+				"contract_id",
+				"event_type",
+				"event_data",
+				"ledger_close_time",
+				"transaction_hash",
+				"topic1", "topic2", "topic3", "topic4",
+			)
+
+		for index, e := range txEvents {
+			var contractID []byte
+			if e.Event.ContractId != nil {
+				contractID = e.Event.ContractId[:]
+			}
+
+			id := Cursor{Ledger: lcm.LedgerSequence(), Tx: tx.Index, Op: 0, Event: uint32(index)}.String()
+			eventBlob, err := e.MarshalBinary()
+			if err != nil {
+				return err
+			}
+
+			v0, ok := e.Event.Body.GetV0()
+			if !ok {
+				return errors.New("unknown event version")
+			}
+
+			// Encode the topics
+			topicList := make([][]byte, MaxTopicCount)
+			for index, segment := range v0.Topics {
+				seg, err := segment.MarshalBinary()
+				if err != nil {
+					return err
+				}
+				topicList[index] = seg
+			}
+
+			query = query.Values(
+				id,
+				contractID,
+				int(e.Event.Type),
+				eventBlob,
+				lcm.LedgerCloseTime(),
+				transactionHash,
+				topicList[0], topicList[1], topicList[2], topicList[3],
+			)
+		}
+
+		_, err = query.RunWith(eventHandler.stmtCache).Exec()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type ScanFunction func(
+	event xdr.DiagnosticEvent,
+	cursor Cursor,
+	ledgerCloseTimestamp int64,
+	txHash *xdr.Hash,
+) bool
+
+// trimEvents removes all Events which fall outside the ledger retention window.
+func (eventHandler *eventHandler) trimEvents(latestLedgerSeq uint32, retentionWindow uint32) error {
+	if latestLedgerSeq+1 <= retentionWindow {
+		return nil
+	}
+	cutoff := latestLedgerSeq + 1 - retentionWindow
+	id := Cursor{Ledger: cutoff}.String()
+
+	_, err := sq.StatementBuilder.
+		RunWith(eventHandler.stmtCache).
+		Delete(eventTableName).
+		Where(sq.Lt{"id": id}).
+		Exec()
+	return err
```

Comment:
Does this run on every ledger? We might want to be a little more defensive and make this `id <= cutoff`. But maybe that has a performance impact. If we're doing the same thing everywhere then this is fine.

Reply:
Yes run on every Commit of new Ledger. Yep similar to how we trim in Txns